ssz_serialization.nim

```diff
@@ -162,11 +162,36 @@ proc writeVarSizeType(w: var SszWriter, value: auto) {.raises: [Defect, IOError]
     # to internally match the binary representation of SSZ BitLists in memory.
     writeSeq(w, bytes value)
   elif value is object|tuple|array:
-    trs "WRITING OBJECT OR ARRAY"
-    var ctx = beginRecord(w, type value)
-    enumerateSubFields(value, field):
-      writeField w, ctx, astToStr(field), field
-    endRecord w, ctx
+    when isCaseObject(type(value)):
+      trs "WRITING SSZ Union"
+      # TODO:
+      # - Check for enum size range (< 127 or at least max 1 byte sizeof)
+      # - SSZ Union: "Must have at least 1 type option."
+      # - SSZ Union: "Must have at least 2 type options if the first is None"
+      # - SSZ Union: Empty case (No fields) only as first (0) discriminator
+
+      # toSszType for enum is kept local here as we don't want enums to
+      # serialize in general, only for object variants.
+      # TODO: This is not sufficiant as it will still allow to parse enum
+      # fields in the object variant itself. Will probably need some macro
+      # which enumerates the fields instead to take specific action on the
+      # discriminator.
+      template toSszType[E: enum](x: E): uint8 =
+        uint8(x)
+
+      enumerateSubFields(value, field):
+        type T = type toSszType(field)
+
+        when isFixedSize(T):
```

Comment:
There is an assumption here that there is a single field in each case branch, otherwise the code wouldn't be correct as far as I can tell (it wouldn't match the behavior described in the spec).

The assumption and the tricky logic why `enumerateSubFields` gets the job done can be explained in a comment above the "loop".

Reply:
Yup, same comment as above.